### DISCUSSION TITLE
How to setup redux or redux toolkit ?

### DISCUSSION THREAD

**Discussion by JustFly1984** [vikejs/vike#1919] [Discussion]
I have several apps made in gatsby.js, with redux and intl
I've managed to refactor some of them to vite, but some of them I want to refactor to vike, cos it supports ssg and looks like next step after remix.

**Comment by brillout** [comment]
https://vike.dev/redux although I think the docs could be improved. Let me have a look at how/if it can be improved.

**Reply by brillout** [reply]
[Done](https://github.com/vikejs/vike/commit/12f3215c65cb1cee819fee1e0e33e2b890501640). See updated docs, most notably https://vike.dev/stores#ssr.

**Reply by JustFly1984** [reply]
@brillout, thank you for reponse.
The thing is that I've moved from gatsby.js to vite/remix, specifically cos I need to deploy to Cloudflare pages, need redux in head component, so Provider does not refreshed on page navigation, and I need i18n support. And specifically, it is required to support Content Security Policy nonce.

The side-effects I've got is that remix+cloudflare combo is SSR, cos it builds [[path]].ts in functions, and this file grows in size with your project, and it is impossible to disable, at least I've tried.

I have requirements for pure SSG, with all the niceties I've described above. 
Gatsby.js sort of had everything, except they broke the CSP nonce support, which was made with non oficial package gatsby-plugin-scp-nonce, and  now it is impossible to use for any of serios projects, due to the lack of maintainance.

@brillout  I was watching vite conf, and the most interesting part was introduction of vike.

It currently looks almost perfect, except 2 crucial details I'm trying to get a grip: 
- redux provider in root component without page navigation
- CSP nonce support for dev and build.

In your https://github.com/brillout/vike-with-redux/blob/main/renderer/%2BonRenderClient.jsx example I see that Provider is used on every page, which is not desired behavior. Gatsby.js' gatsby-ssr API has special hooks to render global root and page components:

```tsx
export const wrapRootElement: GatsbySSR['wrapRootElement'] =
  function wrapRootElement(props: WrapRootElementNodeArgs): JSX.Element {
    return <Provider store={store}>{props.element}</Provider>
  }

export const wrapPageElement: GatsbySSR['wrapPageElement'] =
  function wrapPageElement(
    args: WrapPageElementNodeArgs<Record<string, unknown>, PageContext>
  ): JSX.Element {
    const locale = args.props.pageContext.locale

    intlManager.setMessagesSync(locale)

    return (
      <DarkModeToggleWrapper>
        <IntlComponent locale={locale}>
          {args.element}
        </IntlComponent>
      </DarkModeToggleWrapper>
    )
  }
  ```

Is it possible to expect wrapRootElement kind of API from vike? 

It is also usefull for websockets, cos you can keep websockets provider in root component, so it doesn't re-render on page navigation

**Reply by brillout** [reply]
Regarding wrapping, I ain’t sure what you want but I don’t see a blocker on Vike’s side. In principle, wrapping happens only once per page render.

As for CSP nonce, feel free to elaborate on why it is important to you.

**Reply by JustFly1984** [reply]
@brillout  In my case, I need wrapping happen only per app rendering. It should not happen on navigation, so per page is not an option for me.
It should be the top component, above the router.

**Reply by brillout** [reply]
You can use `globalThis` or `pageContext.isHydration`.

**Reply by JustFly1984** [reply]
how can I render react-redux `<Provider />` with globalThis or pageContext.isHydration ? I need the store to available for every page, and do not want serialisation/de-serialisation happen on page navigation, but at the same time I'm not going to attach store to the window object on the client, since it allows some attack vectors.

CSP nonce is the ultimate way to secure your code from code injections in the browser runtime, kind of protecting the end user from malware browser extensions. Adding nonce dummy string to `<script>` and `<link css>` tags at build time.

I'm using cloudflare pages CDN to host deployed project, and cloudflare workers to attach a middleware to pages routes, so it modifies the headers adding CSP headers, and replacing dummy string with an actual nonce in html in response.

This is important if you care about your users privacy and security, and got tired about MITM attacks injecting their faulty code into your client's browser, and triggering errors or outright crashing your website. I'm noticing weird shit in Sentry logs constantly. Enabling CSP nonce reduces spam errors received by Sentry. And yes, every error logged is eating the Sentry monthly budget.
how can I render react-redux <Provider /> with globalThis or pageContext.isHydration ? I need the store to available for every page, and do not want serialisation/de-serialisation happen on page navigation, but at the same time I'm not going to attach store to the window object on the client, since it allows some attack vectors.

CSP nonce is the ultimate way to secure your code from code injections in the browser runtime, kind of protecting the end user from malware browser extensions. Adding nonce dummy string to `<script>` and `<link css>` tags at build time.

I'm using cloudflare pages CDN to host deployed project, and cloudflare workers to attach a middleware to pages routes, so it modifies the headers adding CSP headers, and replacing dummy string with an actual nonce in html in response.

This is important if you care about your users privacy and security, and got tired about MITM attacks injecting their faulty code into your client's browser, and triggering errors or outright crashing your website. I'm noticing weird shit in Sentry logs constantly. Enabling CSP nonce reduces spam errors received by Sentry. And yes, every error logged is eating the Sentry monthly budget.

**Reply by brillout** [reply]
> how can I render react-redux `<Provider />` with globalThis or pageContext.isHydration ?

It allows you decide whether you initialize only once or for every page. It's seems to be mostly a user-land question unrelated to Vike at this point, but let me know if you've some evidence that something about Vike is blocking you.

As for CSP, there is a [draft PR](https://github.com/vikejs/vike/pull/1645) implementing it. Contribution welcome to revive & finish it.

**Reply by JustFly1984** [reply]
@brillout 
I don't understand the part about pageContext.isHydration

I've looked at your PR for CSP, and left the comment. My idea is that end user will provide the nonce as string, and it will be set as argument on every link and script, generated by vite/vike. Also end-user should be able to add the same nonce for custom scripts and link tags for SSG.

In case of SSR, it should be generated for every request on the backend side, added to CSP header and injected into html to sign every script and link tag.

Both ways there should be default nonce value which is replaced on every request.

**Reply by brillout** [reply]
:+1: Thanks for the CSP insights.

**Reply by JustFly1984** [reply]
@brillout sure, I like what you've created!
